This adds some physical constants

mbkma opened the pull request:
This introduces some physical constants the way pi and e were introduced. To get this working correctly you have to get this patch from me as well #98, otherwise most constants will be shown as 0. The constants are clickable from a drop-down menu:
![Screenshot at 2019-04-23 21-17-33](https://user-images.githubusercontent.com/39454100/56609410-53ea9000-660d-11e9-897a-6a6fd779a597.png)

sc0w commented:
@mbkma 

> To get this working correctly you have to get this patch from me as well #98, otherwise most constants will be shown as 0.


https://github.com/mate-desktop/mate-calc/pull/98 was already merged and it is needed to test this PR, so, please, can you rebase this PR against master?

rbuj commented:
Can you add some unit tests in https://github.com/mate-desktop/mate-calc/blob/master/src/test-mp-equation.c?

lukefromdc commented:
The first commit really should get a better commit message. "This adds some physical constants" or "adding some physical constants doesn't say much about what it's supposed to do.

rbuj commented:
```patch
diff --git a/src/math-buttons.c b/src/math-buttons.c
index 793ad47..539b9ff 100644
--- a/src/math-buttons.c
+++ b/src/math-buttons.c
@@ -1041,28 +1041,28 @@ const_cb(GtkWidget *widget, MathButtons *buttons)
         GtkWidget *menu;
         struct 
         {
-            gchar *name, *constant;
+            gchar *name, *constant, *tooltip;
         } constants[] = 
         {
             { /* Tooltip for the c₀ component button */
-              N_("Velocity of Light"), "c₀" },
+              N_("Velocity of Light"), "c₀", N_("299,792,458 m/s") },
             { /* Tooltip for the μ₀ component button */
-              N_("Magnetic constant"), "μ₀" },
+              N_("Magnetic constant"), "μ₀", N_("1.2566370614×10⁻⁶ N/A²") },
             { /* Tooltip for the ε₀ button */
-              N_("Electric constant"), "ε₀" },
+              N_("Electric constant"), "ε₀", N_("8.85418782×10⁻¹² s⁴A²/m³kg") },
             { /* Tooltip for the G button */
-              N_("Newtonian constant of gravitation"), "G" },
+              N_("Newtonian constant of gravitation"), "G", N_("6.67408×10⁻¹¹ m³/(s²kg)") },
             { /* Tooltip for the h button */
-              N_("Planck constant"), "h" },
+              N_("Planck constant"), "h", N_("6.62607004×10⁻³⁴ m²kg/s") },
             { /* Tooltip for the ｅ button */
-              N_("Elementary charge"), "ｅ" },
+              N_("Elementary charge"), "ｅ", N_("1.6021766208(98)×10⁻¹⁹ C") },
             { /* Tooltip for the mₑ button */
-              N_("Electron mass"), "mₑ" },
+              N_("Electron mass"), "mₑ", N_("9.10938356×10⁻³¹ kg") },
             { /* Tooltip for the mₚ button */
-              N_("Proton mass"), "mₚ" },
+              N_("Proton mass"), "mₚ", N_("1.672621898(21)×10⁻²⁷ kg") },
             { /* Tooltip for the Nₐ button */
-              N_("Avogrado constant"), "Nₐ" },
-            { NULL, NULL }
+              N_("Avogrado constant"), "Nₐ", N_("6.02214086×10²³ mol⁻¹") },
+            { NULL, NULL, NULL }
         };
 
         menu = buttons->priv->const_menu = gtk_menu_new();
@@ -1072,6 +1072,7 @@ const_cb(GtkWidget *widget, MathButtons *buttons)
             GtkWidget *item;
           
             item = gtk_menu_item_new_with_label(_(constants[i].name));
+            gtk_widget_set_tooltip_text(item, _(constants[i].tooltip));
             g_object_set_data(G_OBJECT(item), "const", g_strdup(constants[i].constant));
             gtk_menu_shell_append(GTK_MENU_SHELL(menu), item);
             g_signal_connect(item, "activate", G_CALLBACK(insert_const_cb), buttons);
```

mbkma commented:
ok, thanks! I'm at my brothers pc at the moment and don't get the hang how to push from my account and not from my brothers

mbkma commented:
@lukefromdc  how can I change the commit message now, should I force push to overwrite the commit message?

raveit65 commented:
@mbkma 
> how can I change the commit message now, should I force push to overwrite the commit message?

`git rebase -i <last-commit-id-before-PR>` in terminal.
In opened vi-editor choose `reword` (all is self-explanation)
When all is done do a `git push -f` for updating your githubp repo.
PR will be updated automatic.

mbkma commented:
@rbuj from where did you take the values of the physical constants? I ask, in order to adjust the actual values to the tooltip values (they should be equal) and to cite the source in the commit message. I took mine from https://web.mit.edu/birge/Public/formulas/phys-const.pdf but I don't care which one you choose.

rbuj commented:
@mbkma 
https://en.wikipedia.org/wiki/Physical_constant#Universal_constants

mbkma commented:
I updated this PR now with a better commit message and adjusted physical constant values to the tooltip ones.
@rbuj I am afraid I can't add proper tests for this one, because in the test function the display format is set to fixed, 

`serializer = mp_serializer_new(MP_DISPLAY_FORMAT_FIXED, options.base, 9);`

so most values will become 0:

```
*FAIL: 'μ₀' -> '0.000001257', expected '0.0000012566370614'
*FAIL: 'ε₀' -> '0', expected '0.00000000000885418782'
*FAIL: 'G' -> '0', expected '0.0000000000667408'
*FAIL: 'h' -> '0', expected '0.000000000000000000000000000000000662607004'
*FAIL: 'ｅ' -> '0', expected '0.00000000000000000016021766208'
*FAIL: 'mₑ' -> '0', expected '0.000000000000000000000000000000910938356'
*FAIL: 'mₚ' -> '0', expected '0.000000000000000000000000001672621898'
```

I could of course check it to be zero but I don't know if that is sensible.

rbuj commented:
Use Test to set the trailing digits:
```pach
diff --git a/src/test-mp-equation.c b/src/test-mp-equation.c
index 4404a51..a57c204 100644
--- a/src/test-mp-equation.c
+++ b/src/test-mp-equation.c
@@ -72,7 +72,7 @@ error_code_to_string(MPErrorCode error)
 
 
 static void
-test(char *expression, char *expected, int expected_error)
+Test(char *expression, char *expected, int expected_error, int trailing_digits)
 {
     MPErrorCode error;
     MPNumber result;
@@ -83,7 +83,7 @@ test(char *expression, char *expected, int expected_error)
         char *result_str;
         MpSerializer *serializer;
 
-        serializer = mp_serializer_new(MP_DISPLAY_FORMAT_FIXED, options.base, 9);
+        serializer = mp_serializer_new(MP_DISPLAY_FORMAT_FIXED, options.base, trailing_digits);
         result_str = mp_serializer_to_string(serializer, &result);
         g_object_unref(serializer);
 
@@ -107,6 +107,7 @@ test(char *expression, char *expected, int expected_error)
     }
 }
 
+#define test(X, Y, Z) Test(X, Y, Z, 9)
 
 static int
 do_convert(const MPNumber *x, const char *x_units, const char *z_units, MPNumber *z, void *data)
@@ -315,6 +316,9 @@ test_equations()
     test("π", "3.141592654", 0);
     test("e", "2.718281828", 0);
 
+    /* Physical constants */
+    Test("μ₀", "0.0000012566370614", 0, 16);
+
     test("z=99", "99", 0);
     test("longname=99", "99", 0);
     //test("e=99", "", PARSER_ERR_BUILTIN_VARIABLE);
```

mbkma commented:
@rbuj  thanks, that's handy :)

mbkma commented:
I updated this branch now, because this PR modifies the gui as well, so this is necessary due to #106
